src/osd/ReplicatedPG.cc

```diff
@@ -3904,7 +3904,7 @@ int ReplicatedPG::do_writesame(OpContext *ctx, OSDOp& osd_op)
   }
 
   while (write_length) {
-    write_op.indata.append(osd_op.indata.c_str(), op.writesame.data_length);
+    write_op.indata.append(osd_op.indata);
```

Comment:
Before this line, there is:
   if (op.writesame.data_length != osd_op.indata.length()) {
     derr << "invalid length ws data length " << op.writesame.data_length << " actual len " << osd_op.indata.length() << dendl;
     return -EINVAL;
   }